### PR TITLE
Designer and Teacher's View fixes, saving & loading perspectives in Designer

### PR DESCRIPTION
**Pull request by Polej**
Fix Merge View problems, improve Designer code, enable save & load perspectives in Designer

### CODE REVIEW

File: src/components/Geogebra/GeogebraInterface.js [hunk 1]
```diff
@@ -84,4 +86,9 @@ export default class {
         this.applet.setCaption(objectLabel, caption);
         this.ignoreUpdates = false;
     }
+
+    setPerspective(perspective) {
+        console.log('Setting perspective:', perspective);
```
Comment (IwoHerka) [inline src/components/Geogebra/GeogebraInterface.js]: Use logger.

File: src/components/Geogebra/GeogebraViews.js [hunk 2]
```diff
@@ -90,18 +90,20 @@ export default class {
             }
         });
 
-        api.service('workshops').on('xml-changed', (workshop) => {
+        /* api.service('workshops').on('xml-changed', (workshop) => {
```
Comment (IwoHerka) [inline src/components/Geogebra/GeogebraViews.js:93]: Why do you comment this out?

File: src/components/Pages/Admin/Designer.vue [hunk 3]
```diff
@@ -382,6 +382,11 @@ export default {
                 await this.get(this.selectedConstruction).then((res) => {
                     this.GI.setXML(res.xml);
                     this.GI.registerGlobalListeners();
+                    this.$log.debug(res);
+                    if (res.properties && res.properties.perspectives) {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: One empty line before new block.

File: src/components/Pages/Admin/Designer.vue [hunk 4]
```diff
@@ -382,6 +382,11 @@ export default {
                 await this.get(this.selectedConstruction).then((res) => {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: If `res` a construction? Then use more descriptive name, e.g. `cons`.

File: src/components/Pages/Admin/Designer.vue [hunk 5]
```diff
@@ -452,6 +462,10 @@ export default {
             this.$log.debug('Send');
 
             const xml = this.GI.getXML();
+            const xmlNoConstr = this.produceXMLWithoutConstructionInside(xml);
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: I would prefer to call XML without construction just meta-information.

File: src/components/Pages/Admin/Designer.vue [hunk 6]
```diff
@@ -467,30 +481,19 @@ export default {
...
-            groups out of ${groupsObjects.length} selected`, ((successes === groupsObjects.length) ? 'success' : 'warning'));
+            const properties = { perspectives };
+            await this.sendConstructionToGroups(groupsObjects, xmlNoConstr, properties);
         },
 
         async sendToAll() {
             this.$log.debug('SendToAll');
 
             const xml = this.GI.getXML();
+            const xmlNoConstr = this.produceXMLWithoutConstructionInside(xml);
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: ^

File: src/components/Pages/Admin/Designer.vue [hunk 7]
```diff
@@ -532,34 +543,34 @@ export default {
...
                 } else {
                     this.showToast('Error while creating/updating workshop', 'warning');
                     console.log(error.message);
                 }
-
                 return correct;
             }
         },
 
+        getElementIdToSendIt(groupId, label) {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: I think it's enough just to say `getElementId`. Don't presuppose where this function will be used, that's up to the client.

File: src/components/Pages/Admin/Designer.vue [hunk 8]
```diff
@@ -573,15 +584,94 @@ export default {
...
+            const parser = new DOMParser();
+
+            const xmlDoc = parser.parseFromString(xml, 'text/xml');
+            const construction = xmlDoc.getElementsByTagName('construction')[0];
+
+            xmlDoc.documentElement.removeChild(construction);
+            this.$log.debug(xmlDoc);
+
+            const xmlText = new XMLSerializer().serializeToString(xmlDoc);
+            return xmlText;
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: Inline the previous line.

File: src/components/Pages/Admin/Designer.vue [hunk 9]
```diff
@@ -573,15 +584,94 @@ export default {
                 },
             }]);
             this.$log.debug('groupId', groupId);
+
             await this.addElements(groupId);
         },
+
+        produceXMLWithoutConstructionInside(xml) {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue:602]: I would like to see here at least some docstring.

File: src/components/Pages/Admin/Designer.vue [hunk 10]
```diff
@@ -573,15 +584,94 @@ export default {
...
+            // We sort the views so that we can later send the ordered
+            // arrangement of the different view tabs present
+            // in the activity designer to the students' views
+            const visibleViewsSorted = visibleViews.sort((a, b) => {
+                const aOrder = a.getAttribute('location').replace(/,/g, '');
+                const bOrder = b.getAttribute('location').replace(/,/g, '');
+
+                if (aOrder > bOrder) {
+                    return -1;
+                /* eslint-disable-next-line no-else-return */
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue]: We can disable this opt for linter.

File: src/components/Pages/Admin/Designer.vue [hunk 11]
```diff
@@ -573,15 +584,94 @@ export default {
...
+                } else if (bOrder > aOrder) {
+                    return 1;
+                } else {
+                    return 0;
+                }
+            });
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue:636]: Add one empty line.

File: src/components/Pages/Admin/Designer.vue [hunk 12]
```diff
@@ -573,15 +584,94 @@ export default {
...
+            // The following loop creates a string of the (encoded)
+            // values of the different views present in the
+            // activity designer (to be sent to the students)
+            let perspectivesMapped = '';
+
+            for (let i = 0; i < visibleViewsSorted.length; i += 1) {
+                const id = visibleViewsSorted[i].getAttribute('id');
+                if (id === '1') {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue:649]: One empty line before new block.

File: src/components/Pages/Admin/Designer.vue [hunk 13]
```diff
@@ -573,15 +584,94 @@ export default {
...
+                } else {
+                    return 0;
+                }
+            });
+            // The following loop creates a string of the (encoded)
+            // values of the different views present in the
+            // activity designer (to be sent to the students)
+            let perspectivesMapped = '';
+
+            for (let i = 0; i < visibleViewsSorted.length; i += 1) {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue:642]: Explain in the docstring where did you get this mapping between integers and letters.

File: src/components/Pages/Admin/Designer.vue [hunk 14]
```diff
@@ -573,15 +584,94 @@ export default {
...
+                    perspectivesMapped += 'D';
+                } else if (id === '32') {
+                    perspectivesMapped += 'L';
+                } else if (id === '64') {
+                    perspectivesMapped += 'B';
+                } else if (id === '512') {
+                    perspectivesMapped += 'T';
+                }
+                /*
+                else if (id == '128') {
```
Comment (IwoHerka) [inline src/components/Pages/Admin/Designer.vue:667]: Why do we leave this uncommented?

**Review by IwoHerka — CHANGES_REQUESTED** [review]

File: src/components/Geogebra/GeogebraViews.js (hunk 2 above)
Comment (Polej) [reply src/components/Geogebra/GeogebraViews.js:93]: Loading xml into admin views changes their sizes, it looks ugly. Setting elements should be enough. There's an open question how to handle perspectives in admin views.

File: src/components/Pages/Admin/Designer.vue (hunk 14 above)
Comment (Polej) [reply src/components/Pages/Admin/Designer.vue:667]: Erm, commented? It is taken from old Mathnet, I wanted to signalize that there are other options, which behavior may need further research. However it is highly likely that we don't need these additional mappings.

**Pull request merged**